README.md
py-cnotify
==========

py-cnotify is a Cythonized fork of py-notify by Paul Pogonyshev (http://gna.org/projects/py-notify).


py-notify
=========

Py-notify is a Python package providing tools for implementing Observer programming pattern. These
tools include signals, conditions and variables.

Signals are lists of handlers that are called when signal is emitted. Conditions are basically boolean
variables coupled with a signal that is emitted when condition state changes. They can be combined
using standard logical operators (not, and, etc.) into compound conditions. Variables, unlike conditions,
can hold any Python object, not just booleans, but they cannot be combined.

All three main concepts support deriving by users. Conditions and variables even have derive_type() method
that can be used to subclass them in just a few lines of code, for instance to create a variable that
restricts its value's type.

Latest Py-notify versions can be found at the homepage: http://home.gna.org/py-notify/


Running Standard Tests
======================

Py-notify ships with a set of standard test cases. However, these test cases have not yet been modified
for py-cnotify. If you want to try running them, just execute

 `./run-tests.py`

in the top-level directory of the project.


Installation
============

The usual `./setup.py build; ./setup.py install` or just `./setup.py install` should build and install
py-cnotify. See file INSTALL for details.

You can also build a Debian package from this project. Run `dpkg-buildpackage -B` to build a binary package 
for your host machine. The package has been tested and is working on Ubuntu Precise Pangolin (12.04) and Raring 
Ringtail (13.04, beta release) on both i386 and amd64.


Reporting Bugs
==============

Use the GitHub issues system for bug reporting. Please include the following information in your bug report:
* Bug description and the way to reproduce the bug.
* Exception stack trace, if any.
* System information, most importantly Python and Cython versions.
* Test case if possible.
* Everything else you find relevant.

If the bug is present in Py-notify as well as py-cnotify, you can submit a bug to the original project as well:
  https://gna.org/bugs/?group=py-notify
Before submitting a bug to the original project, be sure to download and test with the latest SVN version
of Py-notify and verify that the bug is still present. Don't forget to check if the bug has been submitted already.


Copyrights and Licenses
=======================

All files from the original Py-notify project are copyrighted by Paul Pogonyshev, except as noted below or
in each specific file.

* File `setup.py` is explicitly put in Public Domain.
* File `benchmark/configobj.py` is copyright (c) 2005-2006 Michael Foord, Nicola Larosa and released under the terms of the BSD License.

All Cython modifications to the original files are copyrighted by Ryan Pessa.

All source files are licensed under the GNU Lesser General Public License version 3 or (at your option) any
later version.


debian/changelog
py-cnotify (0.3.2~raring1) raring; urgency=low

  * build py-cnotify for raring

 -- Ryan Pessa <ryan@essential-elements.net>  Tue, 16 Apr 2013 14:27:33 -0500

py-cnotify (0.3.2~precise1) precise; urgency=low

  * build py-cnotify for precise

 -- Ryan Pessa <ryan@essential-elements.net>  Tue, 16 Apr 2013 14:27:23 -0500

py-cnotify (0.3.2) raring; urgency=low

  * prepare for debian packaging

 -- Ryan Pessa <ryan@essential-elements.net>  Tue, 16 Apr 2013 13:45:48 -0500


cnotify/all.py
# -*- coding: utf-8 -*-

#--------------------------------------------------------------------#
# This file is part of Py-cnotify.                                   #
#                                                                    #
# Copyright (C) 2007, 2008 Paul Pogonyshev.                          #
#                                                                    #
# This library is free software; you can redistribute it and/or      #
# modify it under the terms of the GNU Lesser General Public License #
# as published by the Free Software Foundation; either version 2.1   #
# of the License, or (at your option) any later version.             #
#                                                                    #
# This library is distributed in the hope that it will be useful,    #
# but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  #
# Lesser General Public License for more details.                    #
#                                                                    #
# You should have received a copy of the GNU Lesser General Public   #
# License along with this library; if not, write to the Free         #
# Software Foundation, Inc., 51 Franklin Street, Fifth Floor,        #
# Boston, MA 02110-1301 USA                                          #
#--------------------------------------------------------------------#


"""
An utility module that can be used to conveniently import all package classes at once.  It
would look like this:
    >>> from cnotify.all import *

Whether to use this feature is up to you.
"""

__docformat__ = 'epytext en'


from cnotify.base      import *
from cnotify.bind      import *
from cnotify.condition import *
from cnotify.gc        import *
from cnotify.mediator  import *
from cnotify.signal    import *
from cnotify.utils     import *
from cnotify.variable  import *



# Local variables:
# mode: python
# python-indent: 4
# indent-tabs-mode: nil
# fill-column: 90
# End:


cnotify/__init__.py
# -*- coding: utf-8 -*-

#--------------------------------------------------------------------#
# This file is part of Py-cnotify.                                   #
#                                                                    #
# Copyright (C) 2007, 2008 Paul Pogonyshev.                          #
#                                                                    #
# This library is free software; you can redistribute it and/or      #
# modify it under the terms of the GNU Lesser General Public License #
# as published by the Free Software Foundation; either version 2.1   #
# of the License, or (at your option) any later version.             #
#                                                                    #
# This library is distributed in the hope that it will be useful,    #
# but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  #
# Lesser General Public License for more details.                    #
#                                                                    #
# You should have received a copy of the GNU Lesser General Public   #
# License along with this library; if not, write to the Free         #
# Software Foundation, Inc., 51 Franklin Street, Fifth Floor,        #
# Boston, MA 02110-1301 USA                                          #
#--------------------------------------------------------------------#


"""
cNotify package provides three main concepts: I{L{signals <signal>}}, I{L{conditions
<condition>}} and I{L{variables <variable>}}.  Signals are basically lists of callables
that can be I{emitted} and then will call all contained callables (I{handler} of a signal)
in turn.  Conditions are boolean values complemented with a signal that is emitted when
condition’s I{state} changes.  Variables are akin to conditions but can hold arbitrary
I{values}, not just booleans.  Conditions, unlike variables, can also be combined using
standard logic operators, like negation, conjunction and so on.

All three concepts provide separation between providers (writers, setters) and listeners
(readers, getters) of some entity.  Conditions and variables make the entity explicit—it
is a boolean state for the former and arbitrary Python object for the latter (though
derived variable classes can restrict the set of allowed values.)

Here is a quick example:

    >>> from cnotify.variable import *
    ... name = Variable ()
    ...
    ... import sys
    ... name.changed.connect (
    ...     lambda string: sys.stdout.write ('Hello there, %s!\\n' % string))
    ...
    ... name.value = 'Chuk'

Note that when setting the C{name} variable, you don’t need to know who, if anyone,
listens to changes to it.  Interested parties take care to express their interest
themselves and are informed upon a change automatically.

Here is a little more elaborate example with the same functionality (it requires U{PyGTK
<http://pygtk.org/>}):

    >>> from cnotify.variable import *
    ... import gtk
    ...
    ... name = Variable ()
    ...
    ... def welcome_user (name_string):
    ...     dialog = gtk.MessageDialog (None, 0, gtk.MESSAGE_INFO, gtk.BUTTONS_OK,
    ...                                 'Hello there, %s!' % name_string)
    ...     dialog.run ()
    ...     dialog.destroy ()
    ...
    ... name.changed.connect (welcome_user)
    ...
    ... def set_name_from_entry (entry):
    ...     name.value = entry.get_text ()
    ...
    ... window = gtk.Window ()
    ... window.set_title ('Enter name')
    ...
    ... entry = gtk.Entry ()
    ... entry.show ()
    ... window.add (entry)
    ...
    ... entry.connect ('activate', set_name_from_entry)
    ... window.connect ('destroy', lambda window: gtk.main_quit ())
    ...
    ... window.present ()
    ...
    ... gtk.main ()

Note that C{window} knows absolutely nothing about how changes to C{name} variable are
handled.  If you play with this example, you will notice one thing: pressing C{Enter} in
the main window twice doesn’t pop the welcoming dialog twice.  That is because both
conditions and variables emit their ‘changed’ signal I{only} when their state/value
actually changes, not on every assignment.

Now a final, quite complicated, example introducing conditions and some other features:

    >>> from cnotify.all import *
    ...
    ... pilots = Variable ()
    ... fuel   = Variable ()
    ...
    ... import sys
    ...
    ... pilots.changed.connect (
    ...     lambda pilots: sys.stdout.write ('Pilots are %s\\n' % pilots))
    ... fuel.changed.connect (
    ...     lambda amount: sys.stdout.write ('Got %d litres of fuel\\n' % amount))
    ...
    ... def ready_state_changed (ready):
    ...     if ready:
    ...         sys.stdout.write ('Ready to get off!\\n')
    ...     else:
    ...         sys.stdout.write ('Missing pilots or fuel\\n')
    ...
    ... ready = pilots.is_true () & fuel.predicate (lambda amount: amount > 0)
    ... ready.store (ready_state_changed)
    ...
    ... pilots.value = 'Jen and Jim'
    ... fuel.value   = 500
    ...
    ... fuel.value   = 0

First line of example shows a way to save typing by importing all package contents at
once.  Whether to use this technique is up to you.  Following lines up to C{ready = ...}
should be familiar.

Now let’s consider that assignment closer.  First, C{L{pilots.is_true ()
<variable.AbstractVariable.is_true>}} code creates a condition that is true depending on
C{pilots} value (true for non-empty sequences in our case.)  It is just a convenience
wrapper over C{L{AbstractVariable.predicate <variable.AbstractVariable.predicate>}}
method.  Now, the latter is also used directly in this line of code.  It creates a
condition that is true as long as variable’s value conforms to the passed in predicate.
In particular, C{fuel.predicate (lambda amount: amount > 0)} creates a condition that is
true if C{fuel}’s value is greater than zero.  Predicate conditions will recompute their
state each time variable’s value changes and that’s the point in using them.

Finally, two just constructed conditions are combined into a third condition using ‘and’
operator (C{&}).  This third condition will be true if and only if I{both} its term
conditions are true.  Conditions support four logic operations: negation, conjunction,
disjunction and xoring (with these operators: C{~}, C{&}, C{|} and C{^}.)  In addition,
each condition has C{L{if_else <condition.AbstractCondition.if_else>}} method, which is
much like Python’s C{if} operator.

The next line introduces one more new method: C{L{store
<base.AbstractValueObject.store>}}.  It is really just like connecting its only argument
to the ‘changed’ signal, except that it is also called once with the current state of the
condition (or value of a variable.)

The example should produce this output::

    Missing pilots or fuel
    Pilots are Jen and Jim
    Got 500 litres of fuel
    Ready to get off!
    Got 0 litres of fuel
    Missing pilots or fuel

Notable here is the output from C{ready_state_changed} function.  It is called once at the
beginning from the C{store} method with the state of C{ready} condition (then C{False}.)
Both later calls correspond to changes in C{ready}’s state.  When both C{pilots} and
C{fuel} variables are set, corresponding predicate conditions become true and so does the
C{ready} condition.  However, when one of the predicate conditions becomes false (as the
result of C{fuel} being set to zero), C{ready} turns false again.  Note that
C{ready_state_changed} is not called in between of setting C{pilots} and C{fuel} variable.
C{ready} state is recomputed, but since it remains the same, ‘changed’ signal is not
emitted.

G{packagetree}
"""

__docformat__ = 'epytext en'


# CONFIGURATION

__version__   = '0.3.2'
"""
Version of Py-cnotify, as a string.
"""

version_tuple = (0, 3, 2)
"""
Version of Py-cnotify, as a tuple of integers.  It is guaranteed that version tuples of
later versions will compare greater that those of earlier versions.
"""

# /CONFIGURATION


# Local variables:
# mode: python
# python-indent: 4
# indent-tabs-mode: nil
# fill-column: 90
# End:


cnotify/gc.py
# -*- coding: utf-8 -*-

#--------------------------------------------------------------------#
# This file is part of Py-cnotify.                                   #
#                                                                    #
# Copyright (C) 2007, 2008 Paul Pogonyshev.                          #
#                                                                    #
# This library is free software; you can redistribute it and/or      #
# modify it under the terms of the GNU Lesser General Public License #
# as published by the Free Software Foundation; either version 2.1   #
# of the License, or (at your option) any later version.             #
#                                                                    #
# This library is distributed in the hope that it will be useful,    #
# but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  #
# Lesser General Public License for more details.                    #
#                                                                    #
# You should have received a copy of the GNU Lesser General Public   #
# License along with this library; if not, write to the Free         #
# Software Foundation, Inc., 51 Franklin Street, Fifth Floor,        #
# Boston, MA 02110-1301 USA                                          #
#--------------------------------------------------------------------#


"""
A module for protecting objects from garbage collector.  Sometimes, objects that don’t
have a reference to them (and so are valid garbage collector targets) need to stay alive.
Good example of this are logic L{conditions <condition>}: their state can change because
they have term conditions, yet they may be not referenced from anywhere, since handlers
don’t need a reference to notice a state change.

This module defines both a simple L{interface <AbstractGCProtector>} and several
implementations, some, which are suitable for production use (C{L{FastGCProtector}}), some
for debugging purposes (C{L{RaisingGCProtector}}, C{L{DebugGCProtector}}.)

Py-cnotify classes use value of the C{AbstractGCProtector.default} variable as the
protector instance.  In case you run into a problem, set it to an instance of
C{DebugGCProtector} or a similar class to track the problem down (somewhere near your
program beginning).  However, we believe that Py-cnotify classes must not cause problems
themselves, they may pop up only if you use a garbage-collection protector yourself.
"""

__docformat__ = 'epytext en'
__all__       = ('GCProtectorMeta', 'AbstractGCProtector', 'StandardGCProtector',
                 'SlowGCProtector',
                 'HAVE_FAST_IMPLEMENTATIONS',
                 'FastGCProtector', 'RaisingGCProtector', 'DebugGCProtector')


from cnotify.utils import _PYTHON_IMPLEMENTATION, raise_not_implemented_exception



class GCProtectorMeta (type):

    """
    A meta-class for C{L{AbstractGCProtector}}.  Its only purpose is to define
    C{L{default}} property of the class.  In principle, it can be used for your classes
    too, but better subclass C{AbstractGCProtector} instead.
    """

    def _get_default (self):
        return _default

    def _set_default (self, default):
        global _default
        if default is _default:
            return

        if not isinstance (default, AbstractGCProtector):
            raise ValueError ('can only set AbstractGCProtector.default to an instance of '
                              'AbstractGCProtector; got %s instead' % type (_default))

        try:
            num_active_protections = _default.num_active_protections
        except AttributeError:
            num_active_protections = None

        if num_active_protections:
            raise ValueError ('cannot set a different GC protector: current has active protections '
                              '(num_active_protections = %s)' % num_active_protections)

        _default = default


    default = property (_get_default, _set_default,
                        doc = ("""
                               Current default GC protector.  Starts out as an instance of
                               C{L{FastGCProtector}}, but can be changed for debugging
                               purposes.  Note that setting this class property is only
                               possible if current default protector doesn’t have any
                               active protections, i.e. if its C{num_active_protections}
                               property is zero (or has any false truth value).  This is
                               generally true only at the start of the program, so you
                               cannot arbitrarily switch protectors.  Doing so would lead
                               to unpredictable consequences, up to crashing the
                               interpreter, therefore the restriction.
                               """))


    del _get_default, _set_default



def protect (self, object):
    """
    Protect C{object} from being garbage-collected.  It is legal to protect same C{object}
    several times and an object is prevented from being garbage-collected if it has been
    protected at least once.  As a special case, if C{object} is C{None}, this function
    does nothing.

    For convenience, this function always returns C{object} itself.

    @rtype: C{object}
    """

    raise_not_implemented_exception (self)

def unprotect (self, object):
    """
    Unprotect C{object}.  If has been protected once only or exactly one time more than
    times it has been unprotected, make it a legal target for garbage collection again.
    It is an error to call C{unprotect} more times than C{protect} for a same object, and
    descendant behaviour in this case is undefined.  It may even crash Python.  However,
    as a special case, if C{object} is C{None}, this function does nothing.  In
    particular, it is legal to ‘unprotect’ C{None} without having protected it first,
    because it will be a no-op and not lead to bugs.

    For convenience, this function always returns C{object} itself.

    @rtype: C{object}
    """

    raise_not_implemented_exception (self)

def set_default (self, default):
    """
    This method is deprecated.  Instead, set C{AbstractGCProtector.default} directly.
    """

    AbstractGCProtector.default = default



# This weird class creation only to workaround differences in specifying metaclass between
# Python 2.x and 3.x.
AbstractGCProtector = GCProtectorMeta ('AbstractGCProtector', (object,),
                                       { 'protect':     protect,
                                         'unprotect':   unprotect,
                                         'set_default': set_default })

AbstractGCProtector.__doc__ = \
"""
Simple protector interface with two methods for implementations to define.
"""

del protect, unprotect, set_default



class SlowGCProtector (AbstractGCProtector):

    def __init__(self):
        self.__protected_objects = { }


    def protect (self, object):
        if object is not None:
            object_id         = id (object)
            protected_objects = self.__protected_objects
            protection_data   = protected_objects.get (object_id)

            if protection_data is None:
                protected_objects[object_id] = (object, 1)
            else:
                protected_objects[object_id] = (object, protection_data[1] + 1)

        return object

    def unprotect (self, object):
        if object is not None:
            object_id         = id (object)
            protected_objects = self.__protected_objects
            protection_data   = protected_objects.get (object_id)

            if protection_data is not None:
                num_object_protections = protection_data[1]
                if num_object_protections == 1:
                    del protected_objects[object_id]
                else:
                    protected_objects[object_id] = (object, num_object_protections - 1)
            else:
                raise UnprotectionError ('object is not protected by this %s' % type (self))

        return object


    def num_protected_objects (self):
        return len (self.__protected_objects)

    def num_active_protections (self):
        num_active_protections = 0
        for object, num_object_protections in self.__protected_objects.values ():
            num_active_protections += num_object_protections

        return num_active_protections

    num_protected_objects  = property (num_protected_objects)
    num_active_protections = property (num_active_protections)


    def get_num_object_protections (self, object):
        if object is not None:
            protection_data = self.__protected_objects.get (id (object))
            if protection_data is not None:
                return protection_data[1]

        return 0



class UnprotectionError (ValueError):

    """
    Error that is raised by some L{garbage-collection protectors <AbstractGCProtector>}
    when you try to L{unprotect <AbstractGCProtector.unprotect>} an object more times than
    it had been L{protected <AbstractGCProtector.protect>}.  Of the standard protectors
    only C{L{RaisingGCProtector}} ever raises these exceptions.
    """


if _PYTHON_IMPLEMENTATION == 'CPython':
    from cnotify._gc import DebugGCProtector, FastGCProtector, RaisingGCProtector

    StandardGCProtector       = FastGCProtector
    HAVE_FAST_IMPLEMENTATIONS = True

else:
    StandardGCProtector       = SlowGCProtector
    HAVE_FAST_IMPLEMENTATIONS = False


_default = StandardGCProtector ()



# Local variables:
# mode: python
# python-indent: 4
# indent-tabs-mode: nil
# fill-column: 90
# End:
